lib/Conversion/CoreToFSM/CoreToFSM.cpp

```diff
@@ -0,0 +1,988 @@
+  auto newOutput = opBuilder.create<hw::OutputOp>(output.getLoc(), values);
+  output.erase();
+  FrozenRewritePatternSet frozenPatterns = loadPatterns(*moduleOp.getContext());
+
+  SmallVector<Operation *> opsToProcess;
+  clonedBody.walk([&](Operation *op) { opsToProcess.push_back(op); });
+
+  bool changed = false;
+  GreedyRewriteConfig config;
+  LogicalResult converged =
+      applyOpPatternsGreedily(opsToProcess, frozenPatterns, config, &changed);
+
+  llvm::SmallVector<llvm::DenseSet<size_t>> pv;
+  for (size_t j = 0; j < newOutput.getNumOperands(); j++) {
+    llvm::DenseSet<size_t> possibleValues;
+
+    Value v = newOutput.getOperand(j);
+    addPossibleValues(possibleValues, v);
+    pv.push_back(possibleValues);
+  }
+  std::set<llvm::SmallVector<size_t>> flipped = calculateCartesianProduct(pv);
+  for (llvm::SmallVector<size_t> v : flipped) {
+    llvm::DenseMap<Value, int> m;
+    for (size_t k = 0; k < v.size(); k++) {
+      seq::CompRegOp r = registers[k];
+      m[r] = v[k];
+    }
+
+    int i = regMapToInt(registers, m);
+    visitableStates.insert(i);
+  }
+
+  clonedBody.erase();
+};
+
+// A converter class to handle the logic of converting a single hw.module.
+class HWModuleOpConverter {
+public:
+  HWModuleOpConverter(OpBuilder &builder, HWModuleOp moduleOp,
+                      ArrayRef<std::string> stateRegNames)
+      : moduleOp(moduleOp), opBuilder(builder), stateRegNames(stateRegNames) {}
+  LogicalResult run() {
+    SmallVector<seq::CompRegOp> stateRegs;
+    SmallVector<seq::CompRegOp> variableRegs;
+    WalkResult walkResult = moduleOp.walk([&](seq::CompRegOp reg) {
+      // Check that the register type is an integer.
+      if (!isa<IntegerType>(reg.getType())) {
+        reg.emitError("FSM extraction only supports integer-typed registers");
+        return WalkResult::interrupt();
+      }
+      if (isStateRegister(reg)) {
+        stateRegs.push_back(reg);
+      } else {
+        variableRegs.push_back(reg);
+      }
+      return WalkResult::advance();
+    });
+    if (walkResult.wasInterrupted())
+      return failure();
+    if (stateRegs.empty()) {
+      emitError(moduleOp.getLoc())
+          << "Cannot find state register in this FSM. Use the state-regs "
+             "option to specify which registers are state registers.";
+      return failure();
+    }
+    llvm::DenseMap<Value, size_t> regToIndexMap;
+    int regIndex = 0;
+    SmallVector<seq::CompRegOp> registers;
+    for (seq::CompRegOp c : stateRegs) {
+      regToIndexMap[c] = regIndex;
+      regIndex++;
+      registers.push_back(c);
+    }
+
+    llvm::DenseMap<size_t, StateOp> stateToStateOp;
+    llvm::DenseMap<StateOp, size_t> stateOpToState;
+    // Collect async reset arguments to exclude from the FSM's function type.
+    // FSMs in CIRCT don't have an async reset concept, so these signals are
+    // not passed through. The reset behavior is captured in the initial state.
```

Comment:
Just noticed playing around with this pass - this probably isn't something we need to warn users about since CompReg resets aren't explicitly synchronous (as I recently discovered, we don't technically enforce a reset behaviour over them) so we should be fine to implicitly assume they're synchronous. Thanks for having thought of it and having made the test though!